### PR TITLE
CORTX-32736: UTILS - Consul Support for GCONF

### PR DESCRIPTION
**Pull request by rahul-tripathi-git**
Signed-off-by: Rahul Tripathi <rahul.tripathi@seagate.com>

# Problem Statement
- Support fo Gconf migration to consul backend

# Design
-  Only leaf key allows get call for consul confstore backend.

# Coding 
-  Coding conventions are followed and code is consistent [Y/N]: Y
-  Confirm All CODACY errors are resolved [Y/N]: Y

# Testing
- Are test cases updated along with code changes due to Enhancements/Bugs [Y/N]: NA
- Confirm that new test cases are added to regression and sanity plan files and relevant feature plan files [Y/N]:NA
- Confirm that Test Cases are added for new features added [Y/N]: NA
- Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability [Y/N]: NA
- Confirm Testing was performed with installed RPM/K8s deployment [Y/N]:  Y

# Review Checklist 
####  Before posting the PR please ensure:
- [x] PR is self reviewed
- [ ] Is there a change in filename/package/module or signature [Y/N]: 
- [ ] If yes for above point, Is a notification sent to all other cortx components [Y/N]
- [ ] New package/s added to setup.py?
- [x] Jira is updated
- [x] Check if the description is clear and explained. 
- [x] Check Acceptance Criterion is defined. 
- [x] All the tests performed should be mentioned before Resolving a JIRA. 
- [x] Verification needs to be done before marked as Closed/Verified.

# KVstore/Confstore feature (changes/fixes) checklist
#### Confirm changes are made for: N/A
- [ ] ConfStore
- [ ] KVStore
- [ ] ConfCli
- [ ] Test cases added for all above 3
- [ ] changes done in all the KVpayloads (ConsulKvPayload, IniKvPayload, KvPayload)

# Documentation
- [ ] Changes done to WIKI / Confluence page: N/A

### CODE REVIEW

**Comment by rahul-tripathi-git** [comment]
retest this please.

File: py-utils/src/setup/utils.py [hunk 1]
```diff
@@ -41,12 +41,11 @@ class Utils:
...
     def _copy_cluster_map(config_path: str):
         Conf.load('cluster', config_path, skip_reload=True)
-        cluster_data = Conf.get('cluster', 'node')
-        for _, node_data in cluster_data.items():
-            hostname = node_data.get('hostname')
-            node_name = node_data.get('name')
-            Conf.set('cluster', f'cluster>{node_name}', hostname)
-        Conf.save('cluster')
+        node_keys = Conf.get_keys('cluster', starts_with='node')
+        hostname_keys = [x for x in node_keys if x.endswith('>hostname')]
```
Comment (cdeshmukh) [inline py-utils/src/setup/utils.py:45]: seems keys with delimeter are hardcoded all over, should have been constants, so changing those keys or delimeter would be easy later, if need be.
No point in restricting just here now, such changes will have to be targeted separately

File: py-utils/test/consul_service/test_consul.py [hunk 2]
```diff
@@ -44,8 +55,8 @@ def test_delete(self):
         result = self.consul.kv.delete(f"test/{Conf.machine_id}")
         self.assertIs(result, True)
 
-    def test_service_running(self):
-        ServiceV().validate('isrunning', ['consul'])
+#    def test_service_running(self):
```
Comment (cdeshmukh) [inline py-utils/test/consul_service/test_consul.py:58]: why disabled?

File: py-utils/test/consul_service/test_consul.py [hunk 3]
```diff
@@ -26,7 +28,16 @@ class TestConsul(unittest.TestCase):
 
     @classmethod
     def setUpClass(cls) -> None:
-        cls.consul = Consul()
+        Conf.load("index", "consul://cortx-consul-server:8500/conf", skip_reload=True)
```
Comment (cdeshmukh) [inline py-utils/test/consul_service/test_consul.py:31]: plz maintain this as constant in same file here and explain why this GConf reference is used like this here for test

File: py-utils/test/consul_service/test_consul.py (hunk 2 above)
Comment (rahul-tripathi-git) [reply py-utils/test/consul_service/test_consul.py]: Enabled this again @cdeshmukh. With changes in Gconf, it was failing initially.

File: py-utils/test/consul_service/test_consul.py (hunk 3 above)
Comment (rahul-tripathi-git) [reply py-utils/test/consul_service/test_consul.py]: Have added a comment in code itself, mentioed the hard coding of Gconf address and the reason behind it.

File: py-utils/src/setup/utils.py (hunk 1 above)
Comment (rahul-tripathi-git) [reply py-utils/src/setup/utils.py:45]: I can use Conf.delim to  get the delimiter but the keys or sub keys needs to be hardcoded, either here within code or a common const file.

File: py-utils/test/consul_service/test_consul.py [hunk 4]
```diff
@@ -17,16 +17,32 @@
...
 from cortx.utils.conf_store import Conf
 from cortx.utils.validator.v_service import ServiceV
 
 
 class TestConsul(unittest.TestCase):
 
+# Hardcoded the Gconf in here because there is no way to get Gconf address.
+#  added this as a class constant. This will enable to run this module independently.
+
+    Gconf = "consul://cortx-consul-server:8500/conf"
```
Comment (cdeshmukh) [inline py-utils/test/consul_service/test_consul.py:32]: make GConf in capital, GCONF

File: py-utils/test/consul_service/test_consul.py [hunk 5]
```diff
@@ -17,16 +17,32 @@
...
+from urllib.parse import urlparse
 from consul import Consul
+
 from cortx.utils.conf_store import Conf
 from cortx.utils.validator.v_service import ServiceV
 
 
 class TestConsul(unittest.TestCase):
 
+# Hardcoded the Gconf in here because there is no way to get Gconf address.
```
Comment (cdeshmukh) [inline py-utils/test/consul_service/test_consul.py:29]: \# GConf url is supplied by CORTX provisioner during mini-provisioning to component mini provisioning interfaces, which is not available for tests here, so hardcoded Gconf address.

File: py-utils/test/consul_service/test_consul.py (hunk 4 above)
Comment (rahul-tripathi-git) [reply py-utils/test/consul_service/test_consul.py]: Done

File: py-utils/src/setup/utils.py (hunk 1 above)
Comment (cdeshmukh) [reply py-utils/src/setup/utils.py:45]: yeah it needs to be done all over utils code as i see.

File: py-utils/test/consul_service/test_consul.py (hunk 5 above)
Comment (rahul-tripathi-git) [reply py-utils/test/consul_service/test_consul.py:29]: Yes Correct, we need this file to be able to run independently hence hardcoded.

File: py-utils/src/utils/common/dbconf.py [hunk 6]
```diff
@@ -41,9 +41,10 @@ def init(cls, cluster_conf):
     @classmethod
     def _get_consul_netloc(cls):
         """Get Consul KV netlocation."""
-        eps = Conf.get(cls._cluster_index, f'cortx>external>consul>endpoints')
-        for ep in eps:
-            parsed_url = urlparse(ep)
+        num_consul_endpoints = Conf.get(cls._cluster_index, f'cortx>external>consul>num_endpoints')
```
Comment (rohit-k-dwivedi) [inline py-utils/src/utils/common/dbconf.py:44]: fstring not needed here

**Review by sachitanands — APPROVED** [review]

File: py-utils/test/consul_service/test_consul.py (hunk 5 above)
Comment (cdeshmukh) [reply py-utils/test/consul_service/test_consul.py:29]: Update above comment with given text i mean, which explains the scenario more clearly

File: py-utils/src/setup/utils.py [hunk 7]
```diff
@@ -41,12 +41,11 @@ class Utils:
     @staticmethod
     def _copy_cluster_map(config_path: str):
         Conf.load('cluster', config_path, skip_reload=True)
-        cluster_data = Conf.get('cluster', 'node')
-        for _, node_data in cluster_data.items():
-            hostname = node_data.get('hostname')
-            node_name = node_data.get('name')
-            Conf.set('cluster', f'cluster>{node_name}', hostname)
-        Conf.save('cluster')
```
Comment (rohit-k-dwivedi) [inline py-utils/src/setup/utils.py:49]: why are we skipping `Conf.save('cluster')`

Comment (rahul-tripathi-git) [reply py-utils/src/setup/utils.py:49]: Conf.save is not needed in Consul based confstore backend. Every set command is directly sent to consul kv.

File: py-utils/src/utils/common/dbconf.py (hunk 6 above)
Comment (rahul-tripathi-git) [reply py-utils/src/utils/common/dbconf.py:44]: Done.

File: py-utils/test/consul_service/test_consul.py (hunk 5 above)
Comment (rahul-tripathi-git) [reply py-utils/test/consul_service/test_consul.py:29]: Done.

**Review by cdeshmukh — APPROVED** [review]

**Comment by cdeshmukh** [comment]
retest this please

**Review by rohit-k-dwivedi — APPROVED** [review]
LGTM

**Review by sachitanands — APPROVED** [review]
looks good

**Comment by cdeshmukh** [comment]
image
Previous 1N deployment for this PR was succesfull

**Pull request merged**